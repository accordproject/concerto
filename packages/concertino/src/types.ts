/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
    IDecorator,
    IImport
} from '@accordproject/concerto-types';

/**
 * Metadata for the Concertino format.
 */
export interface Metadata {
  concertinoVersion: string;
  models: Record<string, {
    concertoVersion?: string;
    sourceUri?: string;
    imports?: IImport[],
    decorators?: IDecorator[];
  }>;
}
/**
 * Represents a vocabulary object.
 */
export interface Vocabulary {
  label?: string;
  path?: { [key: string]: string | null };
  additionalTerms?: Record<string, string>;
}

/**
 * Represents a metadata map.
 */
export interface MetadataMap {
  [key: string]: (string | number | boolean | object)[];
}

export type EnumOption = { vocabulary?: Vocabulary; metadata?: MetadataMap }

/**
 * Unique identifier for a declaration.
 */
export interface FullyQualifiedName {
  namespace: string;
  localName: string;
  version: string;
}

/**
 * Represents the Concertino format.
 */
export interface Concertino {
  declarations: Record<string, Declaration>;
  metadata: Metadata;
}

/**
 * Represents a generic declaration.
 */
export type Declaration =
  | ConceptDeclaration
  | EnumDeclaration
  | ScalarDeclaration
  | MapDeclaration;

/**
 * Represents a ConceptDeclaration.
 */
export interface ConceptDeclaration {
  type: 'ConceptDeclaration';
  properties?: Record<string, Property>;
  extends?: string[];
  isAbstract?: boolean;
  vocabulary?: Vocabulary;
  metadata?: MetadataMap;
  prototype?: string;
  name?: FullyQualifiedName;
}

/**
 * Represents an EnumDeclaration.
 */
export interface EnumDeclaration {
  type: 'EnumDeclaration';
  values: Record<string, EnumOption>;
  vocabulary?: Vocabulary;
  metadata?: MetadataMap;
  name?: FullyQualifiedName;
}

export type ScalarType = 'StringScalar' | 'IntegerScalar' | 'BooleanScalar' | 'DoubleScalar' | 'LongScalar' | 'DateTimeScalar';

/**
 * Represents a ScalarDeclaration.
 */
export interface ScalarDeclaration {
  type: ScalarType;
  vocabulary?: Vocabulary;
  metadata?: MetadataMap;
  regex?: string; // For StringScalar
  length?: [number | undefined, number | undefined]; // For StringScalar
  range?: [number | undefined, number | undefined]; // For IntegerScalar, DoubleScalar, LongScalar
  default?: string | number | boolean; // Default value for scalars
  name?: FullyQualifiedName;
}

/**
 * Represents a MapDeclaration.
 */
export interface MapDeclaration {
  type: 'MapDeclaration';
  keyType: string; // e.g., "String", "DateTime"
  valueType: string; // e.g., "String", "Boolean", "com.test.models@1.2.3.Xyc"
  isRelationshipValue?: boolean; // For relationship map values
  vocabulary?: Vocabulary;
  metadata?: MetadataMap;
  name?: FullyQualifiedName;
}

/**
 * Represents a property in a ConceptDeclaration.
 */
export interface Property {
 /**
   * Property identifier, unique within a Concept declaration
   */
  name: string;

  /**
   * Primitive Concerto type, or fully qualified name for Enum declarations,
   * Relationships, Map Declarations & Scalar declarations
   */
  type: string;

  /**
   * Primitive Concerto type for true primitives and scalar declarations
   */
  scalarType?: string;

  /**
   * Opinionated projection of "@Term" and "@Term_XXXXXXX" decorators
   */
  vocabulary?: Vocabulary;

  /**
   * Structural hints, implicitly false when not specified
   */
  isCircular?: boolean;
  isEnum?: boolean;

  /**
   * Modifiers, implicitly false when not specified
   */
  isOptional?: boolean;
  isArray?: boolean;
  isRelationship?: boolean;
  isIdentifier?: boolean;
  inheritedFrom?: string; // For inherited properties

  /**
   * Decorators
   */
  metadata?: MetadataMap;

  /**
   * Metaproperties
   */
  regex?: string; // For String properties
  length?: [number | undefined , number | undefined]; // For String properties
  range?: [number | undefined , number | undefined]; // Only for Integer, Double & Long types
  default?: string | number | boolean; // // Only for Integer, Double & Long types
}
