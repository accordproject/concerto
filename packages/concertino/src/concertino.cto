namespace concertino.metamodel@0.1.0-alpha.2

import concerto.metamodel@1.0.0.{Decorator, Import}

// IMPORTANT NOTE
// This model is not a specification for the Concertino metamodel because Concertino
// does not use $class discriminators. Therefore, this model cannot be used to validate Concertino files
// without a feature such as https://github.com/accordproject/concerto/issues/542
// Nevertheless, this model is useful for understanding the structure of Concertino files and for generating code.

/**
 * Represents the Concertino format.
 */
concept Concertino {
  o Metadata metadata
  o DeclarationMap declarations
}

/**
 * Represents a map of declarations.
 */
map DeclarationMap {
  o String
  o Declaration
}

/**
 * Metadata for the Concertino format.
 */
concept Metadata {
  o String concertinoVersion
  o ModelMetadata[] models
}

/**
 * Represents metadata for a model.
 */
concept ModelMetadata {
  o String namespace
  o String concertoVersion optional
  o String sourceUri optional
  o Import[] imports optional
  o Decorator[] decorators optional
}

/**
 * Represents a unique identifier for a declaration.
 */
concept FullyQualifiedName {
  o String namespace
  o String localName
  o String version
}

/**
 * Represents a generic declaration.
 */
abstract concept Declaration {
  o String name
  o Vocabulary vocabulary optional
  o MetadataMap metadata optional
}

/**
 * Represents a ConceptDeclaration.
 */
concept ConceptDeclaration extends Declaration {
  o Property[] properties
  o String[] extends optional
  o Boolean isAbstract optional
  o Prototype prototype optional
}

/**
 * Indicates a prototype for a ConceptDeclaration.
 */
enum Prototype {
  o AssetDeclaration
  o EventDeclaration
  o TransactionDeclaration
  o ParticipantDeclaration
}

/**
 * Represents an EnumDeclaration.
 */
concept EnumDeclaration extends Declaration {
  o EnumValue[] values
}

/**
 * Represents a ScalarDeclaration.
 */
abstract concept ScalarDeclaration extends Declaration {}

concept StringScalarDeclaration extends ScalarDeclaration {
  o String regex optional // For StringScalar
  o Integer[] length optional length=2 // For StringScalar
  o String default optional // Default value for StringScalar
}
concept IntegerScalarDeclaration extends ScalarDeclaration {
  o Integer[] range optional length=2  // For IntegerScalar
  o Integer default optional // Default value for IntegerScalar
}
concept BooleanScalarDeclaration extends ScalarDeclaration {
  o Boolean default optional // Default value for BooleanScalar
}
concept DoubleScalarDeclaration extends ScalarDeclaration {
  o Double[] range optional length=2  // For DoubleScalar
  o Double default optional // Default value for DoubleScalar
}
concept LongScalarDeclaration extends ScalarDeclaration {
  o Long[] range optional length=2  // For LongScalar
  o Long default optional // Default value for LongScalar
}
concept DateTimeScalarDeclaration extends ScalarDeclaration {
  o String default optional // Default value for DateTimeScalar
}

/**
 * Represents a MapDeclaration.
 */
concept MapDeclaration extends Declaration {
  o String keyType // e.g., "String", "DateTime"
  o String valueType // e.g., "String", "Boolean", "com.test.models@1.2.3.Xyc"
  o Boolean isRelationshipValue optional // For relationship map values
}

/**
 * Represents a property in a ConceptDeclaration.
 */
abstract concept Property {
  o String name
  o String type // Fully qualified type name or primitive type
  o Boolean isArray optional
  o Boolean isOptional optional
  o Boolean isRelationship optional
  o Boolean isIdentifier optional
  o String inheritedFrom optional // For inherited properties
  o Vocabulary vocabulary optional
  o MetadataMap metadata optional
}

concept StringProperty extends Property {
  o String regex optional // For String properties
  o Integer[] length optional // For String properties
  o String default optional // Default value for StringProperty
}
concept IntegerProperty extends Property {
  o Integer[] range optional length=2 // For Integer properties
  o Integer default optional // Default value for IntegerProperty
}
concept BooleanProperty extends Property {
  o Boolean default optional // Default value for BooleanProperty
}
concept DoubleProperty extends Property {
  o Double[] range optional length=2 // For Double properties
  o Double default optional // Default value for DoubleProperty
}
concept LongProperty extends Property {
  o Long[] range optional length=2 // For Long properties
  o Long default optional // Default value for LongProperty
}
concept DateTimeProperty extends Property {
  o String default optional // Default value for DateTimeProperty
}

/**
 * Represents an enum value in an EnumDeclaration.
 */
concept EnumValue {
  o String name
  o Vocabulary vocabulary optional
  o MetadataMap metadata optional
}

/**
 * Represents a vocabulary object.
 */
concept Vocabulary {
  o String label optional
  o AdditionalTermMap additionalTerms optional
}

/**
 * Represents an additional term in a vocabulary.
 */
map AdditionalTermMap {
  o String
  o String
}

/**
 * Represents a metadata map (i.e. non-vocabulary declarations).
 */
map MetadataMap {
  o String
  o SimpleDecorator
}

concept SimpleDecorator {
    o String[] arguments
}
