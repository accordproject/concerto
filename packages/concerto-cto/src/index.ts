/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Concerto CTO concrete syntax module. Concerto is a framework for defining domain
 * specific models.
 *
 * @module concerto-cto
 */

// Exceptions
import ParseException = require('./parseexception');

// Parser
import Parser = require('./parserMain');

// Printer
import Printer = require('./printer');

// External models resolution
import External = require('./external');

export {
    ParseException,
    Parser,
    Printer,
    External
};