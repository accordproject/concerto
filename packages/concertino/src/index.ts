/**
 * Concertino Format Converter Utility
 *
 * This module provides a standardized way to convert between Concerto metamodel and Concertino format.
 */
import { IModels } from '@accordproject/concerto-types';
import { Concertino } from './types';
import { convertToConcertino } from './concertino';
import { convertToMetamodel } from './metamodel';
import Ajv, { ValidateFunction } from 'ajv';
import { readFileSync } from 'fs';
import { join } from 'path';

/**
 * Conversion options for Concertino format.
 */
export interface ConcertinoOptions {
  /**
   * Version of Concertino to use.
   */
  version?: string;
}

const schema = JSON.parse(readFileSync(join(__dirname, 'concertino.schema.json'), 'utf-8'));
const ajv = new Ajv();

/**
 * Concertino utility class for converting between Concerto metamodel and Concertino format.
 */
export class ConcertinoConverter {
    private options: ConcertinoOptions;
    private validate: ValidateFunction;

    /**
     * Creates a new instance of ConcertinoConverter.
     * @param options - Configuration options for the converter.
     */
    constructor(options: ConcertinoOptions = {}) {
        this.options = {
            version: '0.1.0-alpha.3',
            ...options
        };
        this.validate = ajv.compile(schema);
    }

    /**
     * Convert Concerto metamodel to Concertino format.
     * @param metamodel - The Concerto metamodel to convert.
     * @returns The converted Concertino representation.
     */
    public fromConcertoMetamodel(metamodel: IModels): Concertino {
        const concertino = convertToConcertino(metamodel);
        if (this.options.version) {
            concertino.metadata.concertinoVersion = this.options.version;
        }
        return concertino;
    }

    /**
     * Convert Concertino format to Concerto metamodel.
     * @param concertino - The Concertino format to convert.
     * @returns The converted Concerto metamodel.
     */
    public toConcertoMetamodel(concertino: Concertino): IModels {
        return convertToMetamodel(concertino);
    }


    public isValidateStructure(concertino: Concertino): boolean {
        return this.validate(concertino);
    }

}

// Export individual conversion functions for direct use
export { convertToConcertino, convertToMetamodel };

// Export types
export * from './types';
