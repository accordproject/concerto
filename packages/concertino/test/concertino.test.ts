import { readFileSync } from 'fs';
import { join } from 'path';

import { ModelManager } from '@accordproject/concerto-core';
import { Parser } from '@accordproject/concerto-cto';
import { ConcertinoConverter } from '../src/';
import { determineScalarType, dispatchDeclaration, getInheritanceChain, determinePropertyType } from '../src/concertinoSerializer';
import { readdirSync, statSync } from 'fs';
import { IModel, IModels } from '@accordproject/concerto-types';
import { it, expect, describe } from 'vitest';

const loadAllCtoFiles = (dir: string, models: object[]): void => {
    const files = readdirSync(dir);

    files.forEach((file) => {
        const fullPath = join(dir, file);
        const stats = statSync(fullPath);

        if (stats.isDirectory()) {
            loadAllCtoFiles(fullPath, models);
        } else if (file.endsWith('.cto')) {
            const ctoContent = readFileSync(fullPath, 'utf-8');
            try {
                models.push(Parser.parse(ctoContent, fullPath, { skipLocationNodes: true }));
            } catch (error: any) {
                if ('message' in error && !error.message.includes('is already declared')) {
                    // throw error;
                }
                console.warn(`Skipping. Parsing ${file} failed`);
            }
        }
    });
};

describe('concertino roundtripping (sample models)', () => {
    const models: IModel[] = [];

    const modelManager = new ModelManager({
        strict: true,
        importAliasing: true,
        enableMapType: true,
    });
    loadAllCtoFiles(join(__dirname, './cto/'), models);
    modelManager.fromAst({ models });
    const ast: IModels = modelManager.getAst(true);

    ast.models = ast.models.map(model => {
        return modelManager.resolveMetaModel(model) as IModel;
    });

    const converter = new ConcertinoConverter();
    const concertino = converter.fromConcertoMetamodel(ast);
    const metamodel = converter.toConcertoMetamodel(concertino);

    it('should produce valid concertino', () => {
        const isValid = converter.isValid(concertino);
        expect(converter.getValidationErrors()).toBeNull();
        expect(isValid).toBe(true);
        expect(concertino).toMatchSnapshot();
    });

    ast.models.forEach((model: IModel) => {
        it(`should roundtrip ${model.namespace}`, () => {
            const generatedModel = metamodel.models.find((m: IModel) => m.namespace === model.namespace);
            expect(generatedModel).toStrictEqual(model);
        });
    });
});

describe('concertino edge cases and error handling', () => {
    it('should roundtrip sourceUri', () => {
        const converter = new ConcertinoConverter();
        const ast = {
            $class: 'concerto.metamodel@1.0.0.Models',
            models: [{
                $class: 'concerto.metamodel@1.0.0.Model',
                namespace: 'sourceUriTest@1.0.0',
                sourceUri: 'uri'
            }],
        };
        const concertino = converter.fromConcertoMetamodel(ast);
        const metamodel = converter.toConcertoMetamodel(concertino);
        expect(converter.isValid(concertino)).toBe(true);
        expect(concertino).toMatchSnapshot();
        expect(metamodel).toStrictEqual(ast);
    });

    it('should throw for unsupported scalar type', () => {
        expect(() => {
            // @ts-expect-error purposely invalid
            determineScalarType({ $class: 'concerto.metamodel@1.0.0.UnsupportedScalar' });
        }).toThrow(/Unsupported scalar type/);
    });

    it('should throw for unsupported property class', () => {
        expect(() => {
            // @ts-expect-error purposely invalid
            determinePropertyType({ $class: 'concerto.metamodel@1.0.0.UnsupportedProperty' }, { modelNamespace: 'foo' });
        }).toThrow(/Unsupported property class/);
    });

    it('should throw for missing $class in dispatchDeclaration', () => {
        expect(() => {
            dispatchDeclaration({}, { modelNamespace: 'foo' });
        }).toThrow(/Missing \$class property/);
    });

    it('should throw for unsupported object type in dispatchDeclaration', () => {
        expect(() => {
            dispatchDeclaration({ $class: 'concerto.metamodel@1.0.0.Foo' }, { modelNamespace: 'foo' });
        }).toThrow(/Unsupported object type/);
    });

    it('should throw for parent === undefined in getInheritanceChain', () => {
        expect(() => {
            // @ts-expect-error purposely invalid
            getInheritanceChain({ extends: [undefined] }, { declarations: {} });
        }).toThrow(/Parent is undefined/);
    });

    it('should throw for unsupported scalar type', () => {
        const converter = new ConcertinoConverter();
        expect(() => {
            converter.toConcertoMetamodel({
                metadata: {
                    concertinoVersion: '0.0.1-alpha.3',
                    models: {}
                },
                declarations: {
                    'com.test.models@1.2.3.Bar': {
                        'metadata': {
                            // @ts-expect-error purposely invalid
                            'Decorated': [null],
                        },
                        'name': {
                            'localName': 'Bar',
                            'namespace': 'com.test.models',
                            'version': '1.2.3',
                        },
                        'properties': {},
                        'type': 'ConceptDeclaration',
                    }
                },
            });
        }).toThrow(/Unsupported argument type: object/);
    });
});
