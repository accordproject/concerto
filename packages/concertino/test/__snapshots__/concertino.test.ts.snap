// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concertino edge cases and error handling should roundtrip sourceUri 1`] = `
{
  "declarations": {},
  "metadata": {
    "concertinoVersion": "0.1.0-alpha.3",
    "models": {
      "sourceUriTest@1.0.0": {
        "concertoVersion": undefined,
        "decorators": undefined,
        "imports": undefined,
        "sourceUri": "uri",
      },
    },
  },
}
`;

exports[`concertino roundtripping (sample models) should produce valid concertino 1`] = `
{
  "declarations": {
    "com.test.another@1.0.0.Fee": {
      "name": {
        "localName": "Fee",
        "namespace": "com.test.another",
        "version": "1.0.0",
      },
      "regex": "/^[A-Za-z]+$/",
      "type": "StringScalar",
    },
    "com.test.another@1.0.0.Fi": {
      "isAbstract": true,
      "name": {
        "localName": "Fi",
        "namespace": "com.test.another",
        "version": "1.0.0",
      },
      "properties": {
        "someString": {
          "isOptional": true,
          "name": "someString",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "com.test.another@1.0.0.Fo": {
      "extends": [
        "com.test.another@1.0.0.Fi",
      ],
      "name": {
        "localName": "Fo",
        "namespace": "com.test.another",
        "version": "1.0.0",
      },
      "properties": {
        "personId": {
          "isIdentifier": true,
          "name": "personId",
          "scalarType": "String",
          "type": "String",
        },
        "someString": {
          "inheritedFrom": "com.test.another@1.0.0.Fi",
          "isOptional": true,
          "name": "someString",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "com.test.another@1.0.0.Fum": {
      "name": {
        "localName": "Fum",
        "namespace": "com.test.another",
        "version": "1.0.0",
      },
      "properties": {
        "$identifier": {
          "isIdentifier": true,
          "name": "$identifier",
          "type": "String",
        },
        "someString": {
          "name": "someString",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "com.test.models@1.2.3.AuditEvent": {
      "name": {
        "localName": "AuditEvent",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "properties": {
        "$timestamp": {
          "name": "$timestamp",
          "type": "DateTime",
        },
      },
      "prototype": "TransactionDeclaration",
      "type": "ConceptDeclaration",
    },
    "com.test.models@1.2.3.Bar": {
      "metadata": {
        "Decoratored": [],
      },
      "name": {
        "localName": "Bar",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "properties": {
        "someString": {
          "name": "someString",
          "scalarType": "String",
          "type": "String",
          "vocabulary": {
            "label": "some string",
          },
        },
      },
      "type": "ConceptDeclaration",
      "vocabulary": {
        "additionalTerms": {
          "Description": "bar",
        },
        "label": "bar",
      },
    },
    "com.test.models@1.2.3.Dictionary": {
      "keyType": "String",
      "name": {
        "localName": "Dictionary",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "type": "MapDeclaration",
      "valueType": "String",
    },
    "com.test.models@1.2.3.Foo": {
      "name": {
        "localName": "Foo",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "properties": {
        "bar": {
          "name": "bar",
          "type": "com.test.models@1.2.3.Bar",
          "vocabulary": {
            "label": "bar prop",
          },
        },
        "fum": {
          "isArray": true,
          "isOptional": true,
          "isRelationship": true,
          "metadata": {
            "Decorated": [
              "my relationship decorator",
            ],
          },
          "name": "fum",
          "type": "com.test.another@1.0.0.Fum",
          "vocabulary": {
            "additionalTerms": {
              "tooltip": "my relationship tooltip",
            },
            "label": "my relationship",
          },
        },
        "myEnum": {
          "metadata": {
            "Decoratored": [
              1,
              true,
              "a",
            ],
            "Decoratored2": [
              {
                "type": "com.test.models@1.2.3.Bar",
              },
            ],
            "Decoratored3": null,
          },
          "name": "myEnum",
          "type": "com.test.models@1.2.3.MyEnum",
          "vocabulary": {
            "additionalTerms": {
              "Description": "my enum description",
            },
            "label": "my enum",
          },
        },
        "quuux": {
          "default": 10,
          "name": "quuux",
          "range": [
            0,
            100,
          ],
          "scalarType": "com.test.models@1.2.3.Quuux",
          "type": "Integer",
        },
        "quux": {
          "length": [
            1,
            100,
          ],
          "name": "quux",
          "regex": "/^[A-Za-z]+$/",
          "scalarType": "com.test.models@1.2.3.Quux",
          "type": "String",
        },
        "qux": {
          "name": "qux",
          "scalarType": "com.test.models@1.2.3.Qux",
          "type": "String",
        },
        "someBoolean": {
          "name": "someBoolean",
          "scalarType": "Boolean",
          "type": "Boolean",
        },
        "someDouble": {
          "default": 5,
          "name": "someDouble",
          "scalarType": "Double",
          "type": "Double",
          "vocabulary": {
            "label": "some double",
          },
        },
        "someInts": {
          "isArray": true,
          "name": "someInts",
          "scalarType": "Integer",
          "type": "Integer",
        },
        "someLong": {
          "isOptional": true,
          "name": "someLong",
          "scalarType": "Long",
          "type": "Long",
        },
        "someTime": {
          "name": "someTime",
          "scalarType": "DateTime",
          "type": "DateTime",
        },
        "x": {
          "length": [
            0,
            1,
          ],
          "name": "x",
          "regex": "/[A-Z]/",
          "scalarType": "String",
          "type": "String",
        },
        "y": {
          "name": "y",
          "range": [
            0,
            1,
          ],
          "scalarType": "Integer",
          "type": "Integer",
        },
      },
      "type": "ConceptDeclaration",
      "vocabulary": {
        "additionalTerms": {
          "tooltip": "foo tooltip",
        },
        "label": "foo",
      },
    },
    "com.test.models@1.2.3.MyEnum": {
      "metadata": {
        "Decorated": null,
      },
      "name": {
        "localName": "MyEnum",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "type": "EnumDeclaration",
      "values": {
        "FIRST": {
          "vocabulary": {
            "additionalTerms": {
              "Description": "first description",
            },
            "label": "first",
          },
        },
        "SECOND": {
          "metadata": {
            "Decorated": [
              "some value",
            ],
          },
          "vocabulary": {
            "label": "second",
          },
        },
      },
      "vocabulary": {
        "label": null,
      },
    },
    "com.test.models@1.2.3.Quuux": {
      "default": 10,
      "name": {
        "localName": "Quuux",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "range": [
        0,
        100,
      ],
      "type": "IntegerScalar",
    },
    "com.test.models@1.2.3.Quux": {
      "length": [
        1,
        100,
      ],
      "name": {
        "localName": "Quux",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "regex": "/^[A-Za-z]+$/",
      "type": "StringScalar",
    },
    "com.test.models@1.2.3.Qux": {
      "metadata": {
        "Decorated": [],
      },
      "name": {
        "localName": "Qux",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "type": "StringScalar",
      "vocabulary": {
        "label": null,
      },
    },
    "com.test.models@1.2.3.RelationshipMap": {
      "isRelationshipValue": true,
      "keyType": "com.test.models@1.2.3.Qux",
      "metadata": {
        "Decorated": null,
      },
      "name": {
        "localName": "RelationshipMap",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "type": "MapDeclaration",
      "valueType": "com.test.models@1.2.3.Fo",
    },
    "com.test.models@1.2.3.Xyc": {
      "extends": [
        "com.test.another@1.0.0.Fo",
        "com.test.another@1.0.0.Fi",
      ],
      "name": {
        "localName": "Xyc",
        "namespace": "com.test.models",
        "version": "1.2.3",
      },
      "properties": {
        "personId": {
          "inheritedFrom": "com.test.another@1.0.0.Fo",
          "isIdentifier": true,
          "name": "personId",
          "scalarType": "String",
          "type": "String",
        },
        "qux": {
          "name": "qux",
          "scalarType": "com.test.models@1.2.3.Qux",
          "type": "String",
        },
        "someString": {
          "inheritedFrom": "com.test.another@1.0.0.Fi",
          "isOptional": true,
          "name": "someString",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.complex@1.0.0.Department": {
      "name": {
        "localName": "Department",
        "namespace": "org.example.complex",
        "version": "1.0.0",
      },
      "properties": {
        "code": {
          "name": "code",
          "scalarType": "String",
          "type": "String",
        },
        "manager": {
          "name": "manager",
          "type": "org.example.complex@1.0.0.Employee",
        },
        "name": {
          "name": "name",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.complex@1.0.0.Document": {
      "extends": [
        "org.example.complex@1.0.0.DocumentBase",
      ],
      "name": {
        "localName": "Document",
        "namespace": "org.example.complex",
        "version": "1.0.0",
      },
      "properties": {
        "docusignId": {
          "inheritedFrom": "org.example.complex@1.0.0.DocumentBase",
          "name": "docusignId",
          "scalarType": "String",
          "type": "String",
        },
        "expiryDate": {
          "isOptional": true,
          "name": "expiryDate",
          "scalarType": "DateTime",
          "type": "DateTime",
        },
        "id": {
          "isIdentifier": true,
          "name": "id",
          "scalarType": "String",
          "type": "String",
        },
        "issueDate": {
          "name": "issueDate",
          "scalarType": "DateTime",
          "type": "DateTime",
        },
        "number": {
          "name": "number",
          "scalarType": "String",
          "type": "String",
        },
        "type": {
          "name": "type",
          "type": "org.example.complex@1.0.0.DocumentType",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.complex@1.0.0.DocumentBase": {
      "name": {
        "localName": "DocumentBase",
        "namespace": "org.example.complex",
        "version": "1.0.0",
      },
      "properties": {
        "docusignId": {
          "name": "docusignId",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.complex@1.0.0.DocumentType": {
      "name": {
        "localName": "DocumentType",
        "namespace": "org.example.complex",
        "version": "1.0.0",
      },
      "type": "EnumDeclaration",
      "values": {
        "DRIVING_LICENSE": {},
        "NATIONAL_ID": {},
        "PASSPORT": {},
      },
    },
    "org.example.complex@1.0.0.Employee": {
      "name": {
        "localName": "Employee",
        "namespace": "org.example.complex",
        "version": "1.0.0",
      },
      "properties": {
        "department": {
          "name": "department",
          "type": "org.example.complex@1.0.0.Department",
        },
        "employeeId": {
          "name": "employeeId",
          "scalarType": "String",
          "type": "String",
        },
        "identificationDocuments": {
          "isArray": true,
          "name": "identificationDocuments",
          "type": "org.example.complex@1.0.0.Document",
        },
        "personalInfo": {
          "name": "personalInfo",
          "type": "org.example.sample@1.0.0.Person",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.sample@1.0.0.Address": {
      "name": {
        "localName": "Address",
        "namespace": "org.example.sample",
        "version": "1.0.0",
      },
      "properties": {
        "city": {
          "name": "city",
          "scalarType": "String",
          "type": "String",
        },
        "country": {
          "metadata": {
            "Crud": [
              "Create,Read",
            ],
          },
          "name": "country",
          "scalarType": "String",
          "type": "String",
        },
        "postalCode": {
          "name": "postalCode",
          "scalarType": "String",
          "type": "String",
        },
        "street": {
          "metadata": {
            "description": [
              "Street name and number",
            ],
          },
          "name": "street",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.sample@1.0.0.ContactInfo": {
      "name": {
        "localName": "ContactInfo",
        "namespace": "org.example.sample",
        "version": "1.0.0",
      },
      "properties": {
        "type": {
          "name": "type",
          "scalarType": "String",
          "type": "String",
        },
        "value": {
          "name": "value",
          "scalarType": "String",
          "type": "String",
        },
      },
      "type": "ConceptDeclaration",
    },
    "org.example.sample@1.0.0.Person": {
      "name": {
        "localName": "Person",
        "namespace": "org.example.sample",
        "version": "1.0.0",
      },
      "properties": {
        "address": {
          "name": "address",
          "type": "org.example.sample@1.0.0.Address",
        },
        "contacts": {
          "isArray": true,
          "isOptional": true,
          "name": "contacts",
          "type": "org.example.sample@1.0.0.ContactInfo",
        },
        "firstName": {
          "name": "firstName",
          "scalarType": "String",
          "type": "String",
        },
        "lastName": {
          "name": "lastName",
          "scalarType": "String",
          "type": "String",
          "vocabulary": {
            "label": "surname",
          },
        },
      },
      "type": "ConceptDeclaration",
    },
  },
  "metadata": {
    "concertinoVersion": "0.1.0-alpha.3",
    "models": {
      "com.test.another@1.0.0": {
        "concertoVersion": undefined,
        "decorators": [],
        "imports": [],
        "sourceUri": undefined,
      },
      "com.test.models@1.2.3": {
        "concertoVersion": "^3.0.0",
        "decorators": [
          {
            "$class": "concerto.metamodel@1.0.0.Decorator",
            "name": "NamespaceScopedDecorator",
          },
        ],
        "imports": [
          {
            "$class": "concerto.metamodel@1.0.0.ImportTypes",
            "aliasedTypes": [
              {
                "$class": "concerto.metamodel@1.0.0.AliasedType",
                "aliasedName": "Fo",
                "name": "Fo",
              },
            ],
            "namespace": "com.test.another@1.0.0",
            "types": [
              "Fee",
              "Fi",
              "Fo",
              "Fum",
            ],
          },
        ],
        "sourceUri": undefined,
      },
      "org.example.complex@1.0.0": {
        "concertoVersion": undefined,
        "decorators": [],
        "imports": [
          {
            "$class": "concerto.metamodel@1.0.0.ImportTypes",
            "namespace": "org.example.sample@1.0.0",
            "types": [
              "Person",
            ],
          },
        ],
        "sourceUri": undefined,
      },
      "org.example.sample@1.0.0": {
        "concertoVersion": undefined,
        "decorators": [],
        "imports": [],
        "sourceUri": undefined,
      },
    },
  },
}
`;
